compiler/rustc_hir/src/hir.rs

```diff
@@ -2557,6 +2557,22 @@ pub struct OpaqueTy<'hir> {
     /// originating from a trait method. This makes it so that the opaque is
     /// lowered as an associated type.
     pub in_trait: bool,
+    /// List of arguments captured via `impl use<'a, P, ...> Trait` syntax.
+    pub precise_capturing_args: Option<&'hir [PreciseCapturingArg<'hir>]>,
+}
+
+#[derive(Debug, Clone, Copy, HashStable_Generic)]
+pub enum PreciseCapturingArg<'hir> {
+    Lifetime(&'hir Lifetime),
+    /// Non-lifetime argument (type or const)
+    Param(PreciseCapturingNonLifetimeArg),
+}
+
+#[derive(Debug, Clone, Copy, HashStable_Generic)]
+pub struct PreciseCapturingNonLifetimeArg {
```

Comment:
TODO: Document this.

Specifically, we need to have a `hir::Node` for the HIR id that we attach the type/const param resolution to. Lifetimes don't have this problem, and it's actually kind of detrimental to use a custom node type versus just using `hir::Lifetime`, since `resolve_bound_vars` operates on `hir::Lifetime`s.